TSPL.docc/LanguageGuide/Concurrency.md

```diff
@@ -564,33 +637,192 @@ the explicit parent-child relationships between tasks
+like closing network connections and deleting temporary files.
+
+[`Task.checkCancellation()`]: https://developer.apple.com/documentation/swift/task/3814826-checkcancellation
+[`Task.isCancelled`]: https://developer.apple.com/documentation/swift/task/3814832-iscancelled
+
+```
+let photos = await withTaskGroup(of: Data.self) { taskGroup in
+    let photoNames = await listPhotos(inGallery: "Summer Vacation")
+    for name in photoNames {
+        taskGroup.addTaskUnlessCancelled {
+            guard isCancelled == false else { return nil }
+            return await downloadPhoto(named: name)
+        }
+    }
+
+    var results: [Data] = []
+    for await photo in taskGroup {
+        if let photo { results.append(photo) }
+    }
+    return results
+}
+```
+
+<!-- XXX TR:
+Should the listing above add a call to taskGroup.cancelAll()
+to show that part of cancellation?
+If so, what's the best practice for where to put that?
 -->
 
-For more information about task groups,
-see [`TaskGroup`](https://developer.apple.com/documentation/swift/taskgroup).
+The code above makes several changes from the previous version:
+
+- Each task is added using the
+  `TaskGroup.addTaskUnlessCancelled(priority:operation:)` method,
+  to avoid adding new tasks after the task group has been canceled.
+
+- Each task checks for cancellation
+  before starting to download the photo.
+  If it has been canceled, the task returns `nil`.
+
+- At the end,
+  the task group skips `nil` values when collecting the results.
+  Handling cancellation by returning `nil`
+  means the task group can return a partial result ---
+  the photos that were already downloaded at the time of cancellation ---
+  instead of discarding that completed work.
+
+[`TaskGroup.addTaskUnlessCancelled(priority:operation:)`]: https://developer.apple.com/documentation/swift/taskgroup/addtaskunlesscancelled(priority:operation:)
+
+A full version of this code would include clean-up work
+as part of cancellation,
+like deleting partial downloads and closing network connections.
+
+An task that isn't part of a task group can handle cancellation
+using the [`Task.withTaskCancellationHandler(operation:onCancel:)`][] method.
+For example:
+
+[`Task.withTaskCancellationHandler(operation:onCancel:)`]: https://developer.apple.com/documentation/swift/withtaskcancellationhandler(operation:oncancel:)
+
+```swift
+let video = await Task.withTaskCancellationHandler {
+    let file = makeTemporaryFile()
+    renderSlideshow(of: photos, to: file)
+    return file
+} onCancel: {
+    deleteTemporaryFile(file)
+}
+```
 
 <!--
   OUTLINE
 
-  - A task itself doesn't have any concurrency; it does one thing at a time
+  - cancellation propagates (Konrad's example below)
+
+  ::
+
+      let handle = spawnDetached {
+      await withTaskGroup(of: Bool.self) { group in
+          var done = false
+          while done {
+          await group.spawn { Task.isCancelled } // is this child task canceled?
```

Comment:
This seems to use very old API names, so watch out for that:
- `spawnDetached` -> `Task.detached {}`
- `group.spawn` -> `addTask`

Yeah that example is good :)